src/viewProviders/base.ts

```diff
@@ -343,3 +231,142 @@ export abstract class BaseViewProvider<
     return await window.withProgress(progressOptions, task);
   }
 }
+
+/**
+ * Base class for all tree view providers handling a primary resource type and a parent resource.
+ * @template P The type of the "parent" resource that can be "focused" in the view to determine which
+ * resources will be shown. (Example: `KafkaCluster`, `SchemaRegistry`, `FlinkComputePool`)
+ * @template T The primary resource(s) that will be shown in the view.
+ */
+export abstract class ParentedBaseViewProvider<
+    P extends BaseViewProviderData,
+    T extends BaseViewProviderData,
+  >
+  extends BaseViewProvider<T>
+  implements TreeDataProvider<T>, RefreshableTreeViewProvider
+{
+  /**
+   * The focused 'parent' resource instance associated with this provider.
+   *
+   * Examples:
+   * - Topics view: `KafkaCluster`
+   * - Schemas view: `SchemaRegistry`
+   * - Flink Statements view: `FlinkComputePool`
+   * - Flink Artifacts view: `FlinkComputePool`
+   */
+  resource: P | null = null;
+
+  /** The parent {@link Environment} of the focused resource.  */
+  environment: Environment | null = null;
+
+  /**
+   * Optional {@link EventEmitter} to listen for when this view provider's parent
+   * {@linkcode resource} is set/unset. This is used in order to control the tree view description,
+   * context value, and search string updates internally.
+   */
+  parentResourceChangedEmitter?: EventEmitter<P | null>;
+  /** Optional context value to adjust when the parent {@linkcode resource} is set/unset. */
+  parentResourceChangedContextValue?: ContextValues;
+
+  /**
+   * Set the parent resource for this view provider. If being set to what is already set, the
+   * resource will be refreshed.
+   *
+   * @returns A promise that resolves when the resource is set and any reloads are complete.
+   */
+  async setParentResource(resource: P | null): Promise<void> {
+    this.logger.debug(`setParentResource() called, ${resource ? "refreshing" : "resetting"}.`, {
+      resource,
+    });
+
+    if (this.resource !== resource) {
+      this.setSearch(null); // reset search when parent resource changes
+    }
+
+    if (resource) {
+      if (this.parentResourceChangedContextValue) {
+        setContextValue(this.parentResourceChangedContextValue, true);
+      }
+      this.resource = resource;
+      await this.updateTreeViewDescription();
+      await this.refresh();
+    } else {
+      // edging to empty state
+      this.resource = null;
+      await this.refresh();
+    }
+  }
+
+  /** Set up event listeners for this view provider. */
+  protected setEventListeners(): Disposable[] {
```

Comment:
Same question here about `private` vs `protected` for this method

Reply:
Had to move to `protected` to allow `ParentedBaseViewProvider` to override. Can't override `private` methods in TS.